### PR TITLE
Quantity subsampling

### LINKED ISSUE #144: Bootstrapping, subsample quantity.
Reimplement `Quantity.subsample`:
- make it operating on individual chunks without creating a list of selected samples before the chunk processing.
- avoid passing special arguments  

Target usage is [bootstrapping](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)).

This makes #35 deprecated.

### LINKED ISSUE #156: Get rid of Quantity additional_params
@jbrezmorf:  We have to get rid of `additional_params`, seems that the only remaining place is the `subsample`.

### LINKED ISSUE #165: QuantityMean - Sums full of zeros leads to NaN
`QuantityMean._calculate_mean_var` contains a special case:

        if np.isnan(self._mean).all():
            self._mean = []
            self._var = []

Try to get rid of it.

### PR DESCRIPTION
**Pull request by martinspetlik**
resolve #156
resolve #165

### CODE REVIEW

File: src/mlmc/quantity.py [hunk 1]
```diff
@@ -318,30 +303,61 @@ def ne_op(x, y):
...
+        :param chunk: np.ndarray, shape M, N, 2, where N denotes number of samples in chunk
+        :param subsample_params: instance of SubsampleParams class, it has two parameters:
+                        k: number of samples which we want to get from all chunks
+                        n: number of all samples among all chunks
         :return: np.ndarray
         """
-        quantity_storage = self.get_quantity_storage()
-        n_collected = quantity_storage.n_collected()
+        out = []
+        for _ in range(chunk.shape[1]):
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py:318]: I'm not sure that this proposed method leads to the right statistics.

File: src/mlmc/quantity.py [hunk 2]
```diff
@@ -318,30 +303,61 @@ def ne_op(x, y):
...
+            i = random.randint(0, subsample_params.n)
+            if i < subsample_params.k and i < chunk.shape[1]:
+                out.append(chunk[:, i, :])
+                subsample_params.k -= 1
+            subsample_params.n -= 1
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py:323]: Clever, yet modification of the input quantity violates functional programming paradigm, which may block possible parallelization etc. If we adopt chunk iterator, then this information can be part of it. Still not the ideal solution.

File: src/mlmc/quantity.py [hunk 3]
```diff
@@ -565,8 +591,8 @@ def _calculate_mean_var(self):
 
         # Sums full of zeros leads to NaN
         if np.isnan(self._mean).all():
-            self._mean = []
-            self._var = []
+            self._mean = np.array([])
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py]: Why

File: src/mlmc/quantity.py [hunk 4]
```diff
@@ -603,6 +629,8 @@ def _reshape(self, data):
         :return: np.ndarray, reshaped data, the final data shape depends on the particular QType
                              there is currently a reshape for ArrayType only
         """
+        if data.size == 0:
```
Comment (jbrezmorf) [inline src/mlmc/quantity.py]: Is it necessary?

File: src/mlmc/quantity_estimate.py [hunk 5]
```diff
@@ -51,13 +52,18 @@ def estimate_mean(quantity, chunk_size=512000000):
                 # level_chunk is Numpy Array with shape [M, chunk_size, 2]
                 n_samples[level_id] += chunk.shape[1]
                 n_rm_samples[level_id] += n_mask_samples
+
+                # No samples in chunk
+                if chunk.shape[1] == 0:
```
Comment (jbrezmorf) [inline src/mlmc/quantity_estimate.py:61]: That should not happen.

**Review by jbrezmorf — COMMENTED** [review]
The statistical properties of the algorithm must be carefully tested or we should prove it.

File: src/mlmc/quantity_estimate.py (hunk 5 above)
Comment (martinspetlik) [reply src/mlmc/quantity_estimate.py:61]: It could happen. For example, if we are subsampling, a chunk might be empty.

**Pull request merged**